### PR TITLE
3076 admin filter enhancement

### LINKED ISSUE #3076: Admin Filter Enhancements for Data Files Page
**Description:** 
Adds filter to the DAC Data Files page that filters by submission date and includes options for submissions yesterday, today, the past 7 days, the current month, and the current year. #3077 captures a higher lift enhancement to this use case.

**Acceptance Criteria:** 
- [x] Data files page can be filtered by relative dates including:
   - [x] Yesterday
   - [x] Today
   - [x] Past 7 days
   - [x] This month
   - [x] This year
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
- [ ] Implement relative date filter on data files page
- [ ] Tag in UX for a11y review
- [ ] Run Testing Checklist and confirm all tests pass

**Supporting Documentation:**
- [Research Synthesis w/ all DAC issues](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/uAlDVLhlFIbJw6huUS2x/research-drafts/ofa-admin-experience-synthesis)

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
Pull request closes #3076 _

## How to Test
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
2. submit a datafile
3. goto admin page and check data file page. Make sure you can filter the file
4. you can either use one of the DBs from raft environment, or you can simply create a shell, find your submitted datafile using `DataFile.objects.get(id=X)` and then change submission data by using: `df.created_at=datetime.now(tz=timezone.utc)-timedelta(days=2)`

I did change submission date, and tested for different values, and all worked fine.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x]  Data files page can be filtered by relative dates including:
    - [x] Yesterday
    - [x] Today
    - [x] Past 7 days
    - [x] This month
    - [x] This year
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/data_files/admin.py [hunk 1]
```diff
@@ -60,7 +60,44 @@ def data_file_summary(self, obj):
...
+            elif self.value() in ['0', '7']:
+                last_login__lte=datetime.now(tz=timezone.utc) - timedelta(days=int(self.value()))
+                query_set_ids = [df.id for df in queryset if df.created_at >= last_login__lte]
+                return queryset.filter(id__in=query_set_ids)
+            elif self.value() == '30':
+                query_set_ids = [df.id for df in queryset if df.created_at >= this_month]
+                return queryset.filter(id__in=query_set_ids)
+            elif self.value() == '365':
+                query_set_ids = [df.id for df in queryset if df.created_at >= this_year]
+                return queryset.filter(id__in=query_set_ids)
```
Comment (elipe17) [inline tdrs-backend/tdpservice/data_files/admin.py]: Can the logic for all these blocks starting at line 88 through 95 be condensed into a single `elif` and just use a `timedelta`?

Comment (raftmsohani) [reply tdrs-backend/tdpservice/data_files/admin.py]: Unfortunately no! The reason is definition of last month is not 30days delta, but depending on where in that month we are can be  31~59 days

File: tdrs-backend/tdpservice/data_files/admin.py [hunk 2]
```diff
@@ -61,6 +62,47 @@ def data_file_summary(self, obj):
...
+                ('365', 'This year'),
+            ]
+
+        def queryset(self, request, queryset):
+            """Return a queryset."""
+            yesterday = (datetime.now(tz=timezone.utc) - timedelta(days=1)).replace(
+                hour=0, minute=0, second=0, microsecond=0
+                )
+            this_month = datetime.now(tz=timezone.utc).replace(day=1)
+            this_year = datetime.now(tz=timezone.utc).replace(month=1, day=1)
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/data_files/admin.py]: reset the time as well

File: tdrs-backend/tdpservice/data_files/admin.py [hunk 3]
```diff
@@ -61,6 +62,47 @@ def data_file_summary(self, obj):
                            field=f'{df.id}' + ":" + df.get_status(),
                            url=f"{DOMAIN}/admin/parsers/datafilesummary/{df.id}/change/")
 
+    class by_submission_date(admin.SimpleListFilter):
```
Comment (elipe17) [inline tdrs-backend/tdpservice/data_files/admin.py]: Can we change this to use conventional class naming?

File: tdrs-backend/tdpservice/data_files/admin.py (hunk 1 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/data_files/admin.py]: Oh I see now, thank you for clarifying!

File: tdrs-backend/tdpservice/data_files/admin.py [hunk 4]
```diff
@@ -61,6 +62,47 @@ def data_file_summary(self, obj):
...
+                hour=0, minute=0, second=0, microsecond=0
+                )
+            this_month = datetime.now(tz=timezone.utc).replace(day=1)
+            this_year = datetime.now(tz=timezone.utc).replace(month=1, day=1)
+            if self.value() == '1':
+                query_set_ids = [df.id for df in queryset if df.created_at.replace(
+                    hour=0, minute=0, second=0, microsecond=0
+                    ) == yesterday]
+                return queryset.filter(id__in=query_set_ids)
+            elif self.value() in ['0', '7']:
```
Comment (elipe17) [inline tdrs-backend/tdpservice/data_files/admin.py]: Every filter is working correctly except for `Today`. I submitted a single file and then filtered by `Today` and I have an empty queryset.

File: tdrs-backend/tdpservice/data_files/admin.py [hunk 5]
```diff
@@ -61,6 +62,47 @@ def data_file_summary(self, obj):
...
+                last_login__lte = datetime.now(tz=timezone.utc) - timedelta(days=int(self.value()))
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/data_files/admin.py]: is `last_login` relevant?

Comment (raftmsohani) [reply tdrs-backend/tdpservice/data_files/admin.py]: No!

**Review by andrew-jameson — CHANGES_REQUESTED** [review]
Using the today filter results in 0 data files but past 7 days does show it:

```
In [6]: print(alaska.created_at)
2024-08-27 20:03:13.068607+00:00
```

Screenshot 2024-08-27 at 4 10 16 PM
Screenshot 2024-08-27 at 4 10 32 PM
Screenshot 2024-08-27 at 4 11 27 PM

Perhaps we need to update the test for 'today' to include a positive case?

**Review by victoriaatraft — APPROVED** [review]
@raftmsohani LGTM from a11y perspective. I noticed that the titles of the widgets are a little inconsistent with title casing. Can it be changed to: `By submission date` instead of `By Submission Date ` please? Lastly, can we switch Today and Yesterday so the list is in chronological order please? I noticed `By Program Type` is the other inconsistent title. Thank you!

**Review by andrew-jameson — APPROVED** [review]

**Comment by raftmsohani** [comment]
> @raftmsohani LGTM from a11y perspective. I noticed that the titles of the widgets are a little inconsistent with title casing. Can it be changed to: `By submission date` instead of `By Submission Date ` please? Lastly, can we switch Today and Yesterday so the list is in chronological order please? I noticed `By Program Type` is the other inconsistent title. Thank you!

Good call, will do @victoriaatraft !

**Review by elipe17 — APPROVED** [review]
Verified `Today` works now. LGTM

**Review by ADPennington — APPROVED** [review]
@raftmsohani LGTM 🚀  testing notes ⬇️ @klinkoberstar:

_Before (staging)_
df_filters_before

_after (qasp)_
df_filters_after


_after with filter applied (qasp)_
df_filters_after_applied

*also used testing steps to test all the filters (today, yesterday, last 7 days, this month, this year).

**Pull request merged**